abi.html

```diff
@@ -662,40 +677,102 @@ <h3><a href="#pod"> 2.2 POD Data Types </a></h3>
 <a name="member-pointers"></a>
 <h3><a href="#member-pointers"> 2.3 Member Pointers </a></h3>
 
+<h4><a href="#data-member-pointers"> 2.3.1 Data Member Pointers </a></h4>
+
 <p>
-A pointer to data member is an offset from the base
-address of the class object containing it,
-represented as a <code>ptrdiff_t</code>.
-It has the size and alignment attributes of a <code>ptrdiff_t</code>.
-A NULL pointer is represented as -1.
+The basic ABI properties of data member pointer types are those
+of <code>ptrdiff_t</code>.
 
 <p>
-A pointer to member function is a pair <ptr, adj> as follows:
+A data member pointer is represented as the data member's offset in bytes
+from the address point of an object of the base type, as a
+<code>ptrdiff_t</code>.
 
-<dl>
 <p>
-<dt> <code>ptr</code>:
-<dd> For a non-virtual function, this field is a simple function pointer.
-     (Under current base Itanium psABI conventions,
-     that is a pointer to a GP/function address pair.)
-     For a virtual function,
-     it is 1 plus the virtual table offset (in bytes) of the function,
-     represented as a <code>ptrdiff_t</code>.
-     The value zero represents a NULL pointer,
-     independent of the adjustment field value below.
+A null data member pointer is represented as an offset of <code>-1</code>.
+Unfortunately, it is possible to generate an data member pointer with an
+offset of <code>-1</code> using explicit derived-to-base conversions.
+If this is done, implementations following this ABI may misbehave.
+Platforms adopting this ABI without a backward-compatibility concerns
+should consider using <code>PTRDIFF_MIN</code> as the null data member
+pointer instead.
 
 <p>
-<dt> <code>adj</code>:
-<dd> The required adjustment to <i>this</i>,
-     represented as a <code>ptrdiff_t</code>.
-</dl>
+Note that by <code>[dcl.init]</code>, "zero initialization" of a data
+member pointer object stores a null pointer value into it.  Under this
+representation, that value has a non-zero bit representation.  On most
+modern platforms, data member pointers are the only type with this
+property.
+
+<p>
+Base-to-derived and derived-to-base conversions of a non-null data member
+pointer can be performed by adding or subtracting (respectively) the static
+offset of the base within the derived class.  The C++ standard does not
+permit base-to-derived and derived-to-base conversions of member pointers
+to cross a <code>virtual</code> base relationship, and so a static offset
+is always known.
+
+<h4><a href="#member-function-pointers"> 2.3.2 Member Function Pointers </a></h4>
 
 <p>
-It has the size, data size, and alignment
-of a class containing those two members, in that order.
-(For 64-bit Itanium, that will be 16, 16, and 8 bytes respectively.)
+The basic ABI properties of member function pointer types are those of
+the following class:
+
+<pre>
+  // For the type R (C::*)(P...) cv:
+  using FunctionPointerType = R (*)(C cv*, P...);
+
+  struct {
+    FunctionPointerType ptr;
```

Comment:
I thought about it, but I decided that it might cause problems:
- some psABIs might treat such a union differently than a bare function pointer
- `ptrdiff_t` isn't necessarily the same size as a function pointer